packages/foam-core/src/utils/log.ts

```diff
@@ -0,0 +1,89 @@
+export interface ILogger {
+  debug(message?: any, ...params: any[]): void;
+  info(message?: any, ...params: any[]): void;
+  warn(message?: any, ...params: any[]): void;
+  error(message?: any, ...params: any[]): void;
+  getLevel(): LogLevelThreshold;
+  setLevel(level: LogLevelThreshold): void;
+}
+
+export type LogLevel = 'debug' | 'info' | 'warn' | 'error';
+export type LogLevelThreshold = LogLevel | 'off';
+
+export abstract class BaseLogger implements ILogger {
+  private static severity = {
+    debug: 1,
+    info: 2,
+    warn: 3,
+    error: 4,
+  };
+
+  constructor(private level: LogLevelThreshold = 'info') {}
+
+  abstract log(lvl: LogLevel, msg?: any, ...extra: any[]): void;
+
+  doLog(msgLevel: LogLevel, message?: any, ...params: any[]): void {
+    if (this.level === 'off') {
+      return;
+    }
+    if (BaseLogger.severity[msgLevel] >= BaseLogger.severity[this.level]) {
+      this.log(msgLevel, message, ...params);
+    }
+  }
+
+  debug(message?: any, ...params: any[]): void {
+    this.doLog('debug', message, ...params);
+  }
+  info(message?: any, ...params: any[]): void {
+    this.doLog('info', message, ...params);
+  }
+  warn(message?: any, ...params: any[]): void {
+    this.doLog('warn', message, ...params);
+  }
+  error(message?: any, ...params: any[]): void {
+    this.doLog('error', message, ...params);
+  }
+  getLevel(): LogLevelThreshold {
+    return this.level;
+  }
+  setLevel(level: LogLevelThreshold): void {
+    this.level = level;
+  }
+}
+
+export class ConsoleLogger extends BaseLogger {
+  log(level: LogLevel, msg?: string, ...params: any[]): void {
+    console[level](`[${level}] ${msg}`, ...params);
+  }
+}
+
+export class NoOpLogger extends BaseLogger {
+  log(_l: LogLevel, _m?: string, ..._p: any[]): void {}
+}
+
+export class Logger {
```

Comment:
I feel that the setter in this class should only work once and before any logging is done, but I'm not familiar enough with TS to know how to do it

Reply:
I haven't found a reliable way to execute the logging init code before everything else, doing this in JS is really hard.
Happy to go into detail if you are interested as to why.

For this reason, I went for relaxing that constraint (I do agree with you that it would be the better way to enforce consistency).
The other implication was that I couldn't create named loggers, which I liked the idea of but were could still create a stale state (I actually lost a good hour the other day because the default export got me there!).

Interestingly enough I haven't even seen established logging libraries for JS address this 🤷‍♂️

Reply:
You could have a boolean property on the class, `hasLogged` or something to that effect that is set to true on each of the method calls

Now you can guard the `setDefaultLogger` method with this and throw. Does that work?

Reply:
the problem is not detecting the stale state, it is being able to do something about it :)

Reply:
The boolean idea could kinda work but I guess it's "too much" work and we can just expect the developer to not go around and constantly change it

Reply:
oh yes, we wouldn't pass around to plugins the global `Logger` object, but an instance of `ILogger` interface.
only core foam code can call this function, and that is really meant to only be called at bootstrap time, I can't see any other circumstance, so it should be easy to spot a mistake there..